features/323-login-id/type.md

```diff
@@ -0,0 +1,94 @@
+# Login ID Type
+
+## Background
+
+Skygear Auth supports login with login id and password. This document defines the types of login id.
+
+## Validation and Normalization
+
+Validation and normalization are defined base on login id type. When user signup with login ids, Skygear Auth will validate the login ids and reject the signup if it won't pass. After passing the validation, Skygear Auth will store the normalized login id for login.
+
+When user login with login id, login id key is optional. If login id key is provided, Skygear Auth will normalize the input by type, query the principal and perform the login. If login id key is not provided, Skygear Auth will walk through all login id keys, normalize the input and query the principal. If multiple principals are found, the login process will be stopped and error will be returned.
+
+## Types
+
+This section defines the login id types and its validation and normalization rules.
+
+- [Email](#Email)
+- [Username](#Username)
+- [Phone](#Phone)
+- [Raw](#Raw)
+
+### Email
+
+#### Validation
+
+- [RFC 5322 address](https://tools.ietf.org/html/rfc5322#section-3.4.1) (Compulsory)
+
+#### Normalization
+
+- Change domain part to lower case (Compulsory)
+- Change local part to lower case (Configurable, default ON)
+- Remove words after `+` sign in the local part (Configurable, default OFF)
+- Remove all `.` sign in the local part (Configurable, default OFF)
+
+#### Configuration
+```yaml
+user_config:
+  auth:
+    login_id:
+      email:
+        case_sensitive: false
+        ignore_local_part_after_plus_sign: false
+        ignore_dot: false
+    login_id_keys:
+      - type: email
+        key: email
+```
+
+### Username
+
+#### Validation
+
+- Disallow username with reserved keywords in [Django list](https://github.com/ubernostrum/django-registration/blob/31478a8acbf705a654565105c791f1ec4cdbf581/src/django_registration/validators.py#L127) (Configurable, default ON)
+- Disallow username with user defined list (Configurable, default empty list)
+- Disallow non-ASCII username (Configurable, default OFF)
+- Disallow username with confusing homoglyphs (Configurable, default ON)
```

Comment:
I think we don't need this option, allowing confusing homoglyphs is almost certainly insecure.